### PR TITLE
A new runtime primitive for Array.fill

### PR DESCRIPTION
**Pull request by alainfrisch**
(Somehow related to #8714.)

This PR introduces a new runtime primitives which implements (the unsafe version of) `Array.fill`, in order to make this function faster.  The primitive enforces itself the invariants usually managed by `caml_modify`. Performance gains come from:

  - Avoiding the overhead of calling the C primitive for each element.
  - A single test of whether the array is in the minor heap.
  - A single test of whether the new value is a block in the minor heap.
  - Optimization for the case where the new value is equal to an existing one (in which case invariants are automatically satisfied; even if the old value is an old block, no need to call caml_darken -- the same optim could be ported to caml_modify, btw).

The PR also switches to using this optimized `Array.fill` in `Hashtbl.clear` (and its functorized version); and it adds a fast path for when the table is already empty.

Results for the following micro-benchmark:
````ocaml
let () =
  let a = Array.make 128 None in
  (*Gc.compact ();*)
  for _ = 0 to 10000000 do
    Array.fill a 0 128 None
  done
````

If array in minor heap:  trunk 4.0s => PR 0.3s
If array forced in major heap:  trunk 6.0s => PR 1.0s

### CODE REVIEW

**Comment by dra27** [comment]
Bootstrap is required for flexdll bootstrapping (which is the mingw failure). If my bootstrap fu is correct, it's necessary whenever a primitive is added which is used in the stdlib, since the flexdll bootstrap uses `boot/ocamlc` but with a freshly-compiled stdlib.

File: runtime/array.c [hunk 1]
```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
...
+  intnat len = Long_val(v_len);
+  value* fp;
+
+  /* See caml_modify for invariants */
+
+#ifdef FLAT_FLOAT_ARRAY
+  if (Tag_val(array) == Double_array_tag) {
+    double d = Double_val (val);
+    for (; len > 0; len--, ofs++)
+      Store_double_flat_field(array, ofs, d);
```
Comment (dra27) [inline runtime/array.c:610]:
```c
while (len-- > 0)
  Store_double_flat_flat_field(array, ofs++, d);
```

? (I don't like missing `for` clauses, but equally know a lot of people don't like `++` and `--` in expressions!)

File: runtime/array.c [hunk 2]
```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
...
+    return Val_unit;
+  }
+#endif
+  fp = &Field(array, ofs);
+  if (Is_young(array)) {
+    for (; len > 0; len--, fp++) *fp = val;
```
Comment (dra27) [inline runtime/array.c:616]:
```suggestion
    while (len-- > 0) *fp++ = val;
```

File: runtime/array.c [hunk 3]
```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
...
+  } else {
+    CAMLassert(Is_in_heap(fp));
+    int is_val_young_block = Is_block(val) && Is_young(val);
+    for (; len > 0; len--, fp++) {
+      old = *fp;
```
Comment (dra27) [inline runtime/array.c]: I think it's worth an explicit comment that this duplicates the logic in `caml_modify`

File: runtime/array.c [hunk 4]
```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
...
+      if (old == val) continue;
+      *fp = val;
+      if (Is_block(old)) {
+        if (Is_young(old)) continue;
+        if (caml_gc_phase == Phase_mark) caml_darken(old, NULL);
+      }
+      if (is_val_young_block) add_to_ref_table (&caml_ref_table, fp);
+    }
+    /* Should we call caml_check_urgent_gc? */
```
Comment (dra27) [inline runtime/array.c]: Why?

**Review by dra27 — COMMENTED** [review]
I think it's also worth a cross-referencing comment in `memory.c` as well.

I don't think any of the tests exercise arrays in the major heap?

Are there multicore concerns with optimisations like this, @stedolan?

File: runtime/array.c [hunk 5]
```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
...
+    double d = Double_val (val);
+    for (; len > 0; len--, ofs++)
+      Store_double_flat_field(array, ofs, d);
+    return Val_unit;
+  }
+#endif
+  fp = &Field(array, ofs);
+  if (Is_young(array)) {
+    for (; len > 0; len--, fp++) *fp = val;
+  } else {
```
Comment (stedolan) [inline runtime/array.c:617]: The other cases (flat float arrays and minor arrays) are obviously correct and look like clear wins. However, this `else` block duplicates a lot of subtle code from `caml_modify` and is quite hard to follow. (It's no worse than `caml_modify` in this respect, but at least that one is commented).

I think the block could be replaced by the following, and would still optimise the cases advertised in the PR description:

```
for (; len > 0; len--, fp++) {
  if (*fp != val) caml_modify(fp, val);
}
```

**Comment by alainfrisch** [comment]
> I don't think any of the tests exercise arrays in the major heap?

That was the intention behind the call to `Gc.compact` in the new test, at least.

File: runtime/array.c (hunk 4 above)
Comment (alainfrisch) [reply runtime/array.c]: This is what `caml_make_vect`, `caml_make_array`,  `caml_array_blit`, `caml_array_gather` the last two with the comment:
````
  /* Many {caml_modify, caml_initialize} in a row can create a lot of old-to-young refs.
     Give the minor GC a chance to run if it needs to. */
 ````

Comment (dra27) [reply runtime/array.c]: Ah, I see! Makes sense, though I expect @damiendoligez should have the final answer :)

File: runtime/array.c (hunk 5 above)
Comment (alainfrisch) [reply runtime/array.c:617]: The optimizations mentioned in the description (avoiding function calls and repeated `Is_*` tests) also apply the major heap case.

I tested with the following:
````
let () =
  let a = Array.make 128 None in
  Gc.compact ();
  for i = 0 to 10000000 do
    Array.fill a 0 128 (Some i);
  done
````

The simplified version you suggests make this run 2.7x slower then the current PR.  I think we can live with the duplicated logic, considering the impact on performances and the size of the duplicated code (but adding mutual references between the two functions is certainly useful).  Do you agree?

File: runtime/array.c (hunk 3 above)
Comment (alainfrisch) [reply runtime/array.c]: Done.

**Comment by alainfrisch** [comment]
> Bootstrap is required for flexdll bootstrapping (which is the mingw failure). 

Thanks, done!

> I think it's also worth a cross-referencing comment in memory.c as well.

I've added explicit references in both directions.

> (I don't like missing for clauses, but equally know a lot of people don't like ++ and -- in expressions)

Indeed, I like the current version better. But I'll switch if you insist :-)

File: runtime/array.c (hunk 5 above)
Comment (alainfrisch) [reply runtime/array.c:617]: @stedolan : can you confirm if you are ok with the extra bit of code duplication, considering the observed gains?

Comment (stedolan) [reply runtime/array.c:617]: Thanks for the benchmark!

If you think this is an important case to optimise, then I agree this is the right way to do it. Your benchmark shows that you can't get the same level of performance by calling `caml_modify`, and the mutual references in comments give enough pointers for someone else to understand what's going on.

I still can't think of a program that would hit this case a lot, though. (The motivating example of Hashtbl.clear never does). Do you have some program in mind that would be sped up by this, or is this optimisation more for completeness?

Comment (alainfrisch) [reply runtime/array.c:617]: No, honestly, I don't have a program in mind that benefits from this specific case, but it reduces the performance "cliff" between the fast path and the other cases; and addressing it now reduces the administrative overhead of doing it later if it turns out to be useful.

Comment (stedolan) [reply runtime/array.c:617]: OK, makes sense to me!

File: runtime/array.c (hunk 4 above)
Comment (alainfrisch) [reply runtime/array.c]: @damiendoligez, or another GC expert: do you have an opinion on where calling caml_check_urgent_gc is a good idea there?

Comment (damiendoligez) [reply runtime/array.c]: Yes, you should call `caml_check_urgent_gc` here.
```suggestion
    caml_check_urgent_gc (Val_unit);
```

Comment (alainfrisch) [reply runtime/array.c]: Thanks.

Can the call be conditioned to `is_val_young_block` (i.e. `Is_block(val) && Is_young(val)`)?

For future reference, can you explain a bit why this is needed here?  The functions does nothing (except for calling `caml_memprof_handle_postponed`) if neither `caml_requested_major_slice` nor `caml_requested_minor_gc` are set.  But AFAICT, those flags cannot be set during the execution of `caml_array_fill` (the only function calls are `caml_darken` and `add_to_ref_table`, but I could not find a code path from them leading to setting one of the flags).  So this would mean the flags were already set before?

Comment (alainfrisch) [reply runtime/array.c]: Suggestion merged, but still interested in the explanation @damiendoligez

Comment (damiendoligez) [reply runtime/array.c]: Unless I'm mistaken, `add_to_ref_table` can call `realloc_ref_table`, which can call `caml_request_minor_gc`, which sets `caml_requested_minor_gc`.

You don't want to use arbitrary amounts of memory for the remembered set (ref_table) so when it gets too large you trigger a minor GC to eliminate it.

File: stdlib/hashtbl.ml [hunk 6]
```diff
@@ -81,11 +81,11 @@ let create ?(random = !randomized) initial_size =
...
 let clear h =
-  h.size <- 0;
-  let len = Array.length h.data in
-  for i = 0 to len - 1 do
-    h.data.(i) <- Empty
-  done
+  if h.size > 0 then begin
+    h.size <- 0;
+    let len = Array.length h.data in
+    Array.fill h.data 0 len Empty
```
Comment (damiendoligez) [inline stdlib/hashtbl.ml]: Why not simply `Array.fill h.data 0 (Array.length h.data) Empty` ?

**Review by damiendoligez — APPROVED** [review]
A minor nitpick, otherwise this looks good.

File: runtime/array.c (hunk 4 above)
Comment (alainfrisch) [reply runtime/array.c]: Ah, sorry, I don't know how I missed that.  (Side note: shouldn't `caml_request_minor_gc` and `caml_requested_minor_gc`be moved from signals.c to e.g. minor_c?  They are not really signal-related.)

Comment (damiendoligez) [reply runtime/array.c]: In fact, you could do `if (is_val_young_block) caml_check_urgent_gc (Val_unit);`

File: stdlib/hashtbl.ml (hunk 6 above)
Comment (alainfrisch) [reply stdlib/hashtbl.ml]: I just kept that from the original code.  I've now added a commit following your suggestion.

**Comment by alainfrisch** [comment]
Thanks @damiendoligez .  I will merge once the CI is green again.

File: runtime/array.c (hunk 4 above)
Comment (alainfrisch) [reply runtime/array.c]: Ok, done.
